File: python/61. Rotate List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        dummyhead = ListNode(0)
        dummyhead.next = head
        if not head or not head.next:
            return head
        length = 0
        cur = head
        while cur:
            cur = cur.next
            length += 1
        k = k % length
        if k == 0:
            return head
        end = head
        pre = dummyhead
        while k > 0:
            pre = pre.next
            end = end.next
            k -= 1
        start = head
        while end:
            end = end.next
            pre = pre.next
            start = start.next
        cur = head
        while cur.next:
            if cur.next == start:
                cur.next = None
                break
            cur = cur.next
        pre.next= head
        return start

File: cpp/51-100/61. Rotate List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* rotateRight(ListNode* head, int k) {
		if(!head)
			return NULL;
		int length = 0;
		ListNode *ptr = head;
		while(ptr){
			ptr = ptr->next;
			length ++;
		}
		k = k % length;
		ListNode *ptr1, *ptr2;
		ptr1 = ptr2 = head;
		while (k > 0) {
			ptr2 = ptr2->next;
			k--;
		}
		while (ptr2->next) {
			ptr1 = ptr1->next;
			ptr2 = ptr2->next;
		}
		ptr2->next = head;
		head = ptr1->next;
		ptr1->next = NULL;
		return head;

	}
};